### PR TITLE
Template tests

### PR DESCRIPTION
**Pull request by osamaramihafez**
Create test, "uploading a template". Ensure that backend storage is not polluted with test files. Remove created files as soon as the test is over. Fix Mock API to ensure that duplicate copies of a file_name do not exist.

### CODE REVIEW

**Comment by siefkenj** [comment]
PR description?

File: frontend/src/api/mockAPI/contract_templates.js [hunk 1]
```diff
@@ -39,9 +39,18 @@ export class ContractTemplate extends MockAPIController {
             if (message) {
                 throw new Error(message);
             }
+            return;
```
Comment (siefkenj) [inline frontend/src/api/mockAPI/contract_templates.js]: This shouldn't be needed.

File: frontend/src/api/mockAPI/contract_templates.js [hunk 2]
```diff
@@ -39,9 +39,18 @@ export class ContractTemplate extends MockAPIController {
...
         }
     }
     upload({ file_contents, file_name }) {
+        const file_names = Object.keys(
+            this.data.contract_templates_by_filename
+        );
+        if (file_names.includes(file_name)) {
```
Comment (siefkenj) [inline frontend/src/api/mockAPI/contract_templates.js]: You can just do `if (file_name in this.data.contract_templates_by_filename)`

Comment (siefkenj) [reply frontend/src/api/mockAPI/contract_templates.js]: Or maybe more clear `if(this.data.contract_templates_by_filename[file_name] != null)`

File: frontend/src/tests/api.test.js [hunk 3]
```diff
@@ -23,6 +23,7 @@ import { userPermissionsTests } from "./user-permission-tests";
 import { applicantTests } from "./applicant-tests";
 import { instructorsPermissionTests } from "./instructor-permission-test";
 import { ddahTests, ddahsEmailAndDownloadTests } from "./ddah-tests";
+import fs from "fs";
```
Comment (siefkenj) [inline frontend/src/tests/api.test.js]: Put Node imports at the top.

File: package-lock.json [hunk 4]
```diff
@@ -668,6 +668,11 @@
             "resolved": "https://registry.npmjs.org/flatted/-/flatted-2.0.2.tgz",
             "integrity": "sha512-r5wGx7YeOwNWNlCA0wQ86zKyDLMQr+/RB8xy74M4hTphfmjlijTSSXGuH8rnvKZnfT9i+75zmd8jcKdMR4O6jA=="
         },
+        "fs": {
+            "version": "0.0.1-security",
+            "resolved": "https://registry.npmjs.org/fs/-/fs-0.0.1-security.tgz",
```
Comment (siefkenj) [inline package-lock.json]: Did you look at the description of this package? https://www.npmjs.com/package/fs

`fs` should be natively supplied by Node. No need to install it.

File: frontend/src/tests/template-tests.js [hunk 5]
```diff
@@ -210,5 +210,27 @@ export function templatesTests(api) {
...
+        template.content = bytesToBase64(template.content);
+        checkPropTypes(offerTemplatePropTypes, template);
+
+        const resp2 = await apiPOST(
+            `/admin/contract_templates/upload`,
+            template
+        );
+        expect(resp2).toHaveStatus("success");
+
+        // Testing to ensure that duplicate file names are not added (need to fix mock API)
```
Comment (siefkenj) [inline frontend/src/tests/template-tests.js]: Do you still need to fix the mock API?

File: frontend/src/tests/api.test.js [hunk 6]
```diff
@@ -114,7 +119,7 @@ describe("Mock API tests", () => {
         sessionsTests(mockAPI);
     });
     describe("`/admin/sessions/:session_id/contract_templates` tests", () => {
-        templatesTests(mockAPI);
+        templatesTests(mockAPI, () => {});
```
Comment (siefkenj) [inline frontend/src/tests/api.test.js]: Is this change needed?

File: frontend/src/api/mockAPI/contract_templates.js [hunk 7]
```diff
@@ -42,6 +42,15 @@ export class ContractTemplate extends MockAPIController {
         }
     }
     upload({ file_contents, file_name }) {
+        // const file_names = Object.keys(
+        //     this.data.contract_templates_by_filename
+        // );
+        // if (file_names.includes(file_name)) {
```
Comment (siefkenj) [inline frontend/src/api/mockAPI/contract_templates.js]: Why is this comment here?

File: frontend/src/tests/template-tests.js [hunk 8]
```diff
@@ -210,5 +210,27 @@ export function templatesTests(api) {
...
+        );
+        expect(resp2).toHaveStatus("success");
+
+        // Testing to ensure that duplicate file names are not added
+        const resp3 = await apiPOST(
+            `/admin/contract_templates/upload`,
+            template
+        );
+        expect(resp3).toHaveStatus("error");
+        if (callback) callback();
```
Comment (siefkenj) [inline frontend/src/tests/template-tests.js]: Never leave off the `{  }` braces on an `if`

File: frontend/src/tests/template-tests.js [hunk 9]
```diff
@@ -20,7 +20,7 @@ import { base64ToBytes } from "../api/mockAPI/utils";
  * @param {Function} api.apiGET A function that when passed a route will return the get response
  * @param {Function} api.apiPOST A function that when passed a route and data, will return the post response
  */
-export function templatesTests(api) {
+export function templatesTests(api, callback) {
```
Comment (siefkenj) [inline frontend/src/tests/template-tests.js]: Why do you need a callback? Test for and clean up the test file before you rely on its non-existence, instead.

Comment (osamaramihafez) [reply frontend/src/tests/template-tests.js]: I thought it would be cleaner than just leaving the file after the test was over. I can keep the test file though.

File: frontend/src/tests/api.test.js [hunk 10]
```diff
@@ -38,6 +39,15 @@ describe("API tests", () => {
     }, 30000);
 
     describe("template tests", () => {
+        if (
+            fs.existsSync(
+                "/storage_mounted_for_testing/contract_templates/TestTemplate.html"
```
Comment (siefkenj) [inline frontend/src/tests/api.test.js]: Let's store this as a const variable

File: frontend/src/api/mockAPI/contract_templates.js (hunk 7 above)
Comment (siefkenj) [reply frontend/src/api/mockAPI/contract_templates.js]: @osamaramihafez are you going to address this?
